Insert gcode at layer change

wporter82 opened the pull request:
I needed the ability to manually send commands on layer change and couldn't find a solution so I made this one up to fill that need.

Ghostkeeper commented:
As seen on Reddit!

Developers, see issue CURA-5713.

Kriechi commented:
@wporter82 I made a similar script to inject some instructions for timelapse creation with the Duet & RepRapFirmware. I think there is a good overlap - so maybe we can merge our approaches together! Let me know what you think!

wporter82 commented:
@Kriechi I have added the option for parking the head to be optional as well as looked at what your scripts were doing for the Duet. I added the wait for moves to finish but for the picture taking command, the default I have set is to use the Marlin code M240 but can be changed by the user to something like what you have.

LipuFei commented on plugins/PostProcessingPlugin/scripts/TimeLapse.py:
```diff
@@ -0,0 +1,99 @@
+
+        gcode_to_append = self.putValue(M = 400) + ";Wait for moves to finish\n"
+        if park_print_head:
+	        gcode_to_append = self.putValue(G = 90) + ";Absolute positioning\n"
```
This line and the next in this `if` block contain <TAB>s as indents.

alekseisasin commented:
@wporter82 , I have tried _Before_ and _After_ and all of them has the same output, did not see any difference.

alekseisasin commented:
@wporter82 , do you still want to add this script?

wporter82 commented:
I agree with @dector in that this can cause problems if considerations are not taken. I've considered writing a complete plugin just for time lapse instead of going this route of basically find/replace. If this appears to cause more possible problems than it helps solve, I would think it should be abandoned for a better solution.

alekseisasin commented:
Right, a solution which might introduce another issue is not a good scenario.
If will be no more effort done for this ticket then I suggest closing this issue.

diegopradogesto commented on plugins/PostProcessingPlugin/scripts/InsertAtLayerChange.py:
```diff
@@ -0,0 +1,55 @@
+            "version": 2,
+            "settings":
+            {
+                "insert_loc":
```
I would not abbreviate the setting name:
```suggestion
                "insert_location":
```

diegopradogesto commented on plugins/PostProcessingPlugin/scripts/InsertAtLayerChange.py:
```diff
@@ -0,0 +1,55 @@
+
+            if in_layer:
+                index = data.index(layer)
+                if self.getSettingValueByKey("insert_loc") == "before":
```
```suggestion
                if self.getSettingValueByKey("insert_location") == "before":
```

diegopradogesto commented on plugins/PostProcessingPlugin/scripts/InsertAtLayerChange.py:
```diff
@@ -0,0 +1,55 @@
+            else:
+                in_layer = False
+
+            if in_layer:
```
Suggestion: This can be simplified by using `if ";LAYER:" in lines[0]:` and remove the 4 previous layers.

diegopradogesto commented on plugins/PostProcessingPlugin/scripts/TimeLapse.py:
```diff
@@ -0,0 +1,100 @@
+            "version": 2,
+            "settings":
+            {
+                "trigger_cmd":
```
Do not abbreviate setting name:
```suggestion
                "trigger_command":
```

diegopradogesto commented on plugins/PostProcessingPlugin/scripts/TimeLapse.py:
```diff
@@ -0,0 +1,100 @@
+        park_print_head = self.getSettingValueByKey("park_print_head")
+        x_park = self.getSettingValueByKey("head_park_x")
+        y_park = self.getSettingValueByKey("head_park_y")
+        trigger_cmd = self.getSettingValueByKey("trigger_cmd")
```
```suggestion
        trigger_command = self.getSettingValueByKey("trigger_command")
```

diegopradogesto commented on plugins/PostProcessingPlugin/scripts/TimeLapse.py:
```diff
@@ -0,0 +1,100 @@
+        if park_print_head:
+            gcode_to_append += self.putValue(G = 1, F = feed_rate, X = x_park, Y = y_park) + ";Park print head\n"
+        gcode_to_append += self.putValue(M = 400) + ";Wait for moves to finish\n"
+        gcode_to_append += trigger_cmd + ";Snap Photo\n"
```
```suggestion
        gcode_to_append += trigger_command + ";Snap Photo\n"
```

diegopradogesto commented on plugins/PostProcessingPlugin/scripts/TimeLapse.py:
```diff
@@ -0,0 +1,100 @@
+            else:
+                in_layer = False
+
+            if in_layer:
```
Suggestion: This can be simplified by using `if ";LAYER:" in lines[0]:` and remove the 4 previous layers.

diegopradogesto commented:
@wporter82 are you handling the suggested changes? No rush, only to know if we need to keep this in mind or not.

diegopradogesto commented:
I did some changes here 93966281708a1072fe4ae7f4efbfc3e54a179707 and merged it since it was taking a lot of lead time.

Ghostkeeper commented:
We might want to add the option to retract during the move towards the park location as well.
